### PR TITLE
VP-2159 CRUD for product association

### PR DESCRIPTION
**Pull request by AliveMen**
Add ID to ProductAssociation model
Refactoring the UpdateAssociationsAsync method
Remove custom comparer
Refactoring Delete associations method

### CODE REVIEW

File: src/VirtoCommerce.CatalogModule.Data/Services/AssociationService.cs [hunk 1]
```diff
@@ -99,29 +98,29 @@ public async Task SaveChangesAsync(IHasAssociations[] owners)
...
-            var changedEntities = new List<AssociationEntity>();
-
-            var dbAssociations = associations.Select(x => AbstractTypeFactory<AssociationEntity>.TryCreateInstance().FromModel(x)).ToArray();
-
-            changedEntities.AddRange(dbAssociations);
-
             using (var repository = _repositoryFactory())
             {
-                foreach (var changedEntity in changedEntities)
+                var dbExistEntities = await repository.GetAssociationsByIdsAsync(associations.Where(x => !x.IsTransient()).Select(x => x.Id).ToArray());
```
Comment (yecli) [inline src/VirtoCommerce.CatalogModule.Data/Services/AssociationService.cs:103]: please rename to `dbExistentEntities`

File: src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleAssociationsController.cs [hunk 2]
```diff
@@ -55,13 +55,13 @@ public async Task<ActionResult> UpdateAssociations([FromBody] ProductAssociation
...
         /// </summary>
         /// <remarks>Updates the specified association</remarks>
-        /// <param name="ids">associations to delete ids</param>
+        /// <param name="id">association id to delete</param>
         [HttpDelete]
-        [Route("")]
+        [Route("{id}")]
         [Authorize(ModuleConstants.Security.Permissions.Delete)]
-        public async Task<ActionResult> Delete([FromBody]string[] ids)
+        public async Task<ActionResult> Delete(string id)
```
Comment (yecli) [inline src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleAssociationsController.cs:62]: There were the changes in this place in the dev, please rebase to the latest.
Also I think it is not needed to change the API - we should be able to have an ability to delete collection at once.

File: src/VirtoCommerce.CatalogModule.Data/Services/AssociationService.cs (hunk 1 above)
Comment (AliveMen) [reply src/VirtoCommerce.CatalogModule.Data/Services/AssociationService.cs]: renamed

File: src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleAssociationsController.cs (hunk 2 above)
Comment (AliveMen) [reply src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogModuleAssociationsController.cs]: fixed

File: src/VirtoCommerce.CatalogModule.Data/Services/AssociationService.cs [hunk 3]
```diff
@@ -99,29 +98,29 @@ public async Task SaveChangesAsync(IHasAssociations[] owners)
...
-            var changedEntities = new List<AssociationEntity>();
-
-            var dbAssociations = associations.Select(x => AbstractTypeFactory<AssociationEntity>.TryCreateInstance().FromModel(x)).ToArray();
-
-            changedEntities.AddRange(dbAssociations);
-
             using (var repository = _repositoryFactory())
             {
-                foreach (var changedEntity in changedEntities)
+                var dbExistentEntities = await repository.GetAssociationsByIdsAsync(associations.Where(x => !x.IsTransient()).Select(x => x.Id).ToArray());
```
[the remark this answers is not in the document]
Comment (yecli) [reply src/VirtoCommerce.CatalogModule.Data/Services/AssociationService.cs:103]: New `using` syntax is not commonly used in VC projects.
For me personally it is less readable than `using` with the explicit scope.
So I do not think we should change this.

**Comment by AliveMen** [comment]
1. We don't have any association in our sample data. 
2. This PR adds ID field only to DTO object. Database entity already has "ID" field as PK with NOT NULL constraint. 
3. At least before this PR association import hadn't worked at all. 
@tatarincev Is it still breaking changes?

File: src/VirtoCommerce.CatalogModule.Data/Services/AssociationService.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (AliveMen) [reply src/VirtoCommerce.CatalogModule.Data/Services/AssociationService.cs]: done

File: tests/VirtoCommerce.CatalogModule.Tests/ProductAssociationTests.cs [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+                    AssociatedCategoryId = null,
+                    AssociationType = "product",
+                    Priority = 0
+                },
+            },
+        };
+
+        [Theory]
+        [MemberData(nameof(ValidEntities))]
+        public async Task AddNotTransientAssociation(dataModel.AssociationEntity entity)
```
Comment (yecli) [inline tests/VirtoCommerce.CatalogModule.Tests/ProductAssociationTests.cs:43]: Please change all test names according to https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#naming-your-tests, e.g. here

UpdateAssociationsAsync_AddNotTransientAssociation_Added

File: tests/VirtoCommerce.CatalogModule.Tests/ProductAssociationTests.cs [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+                Id = entity.Id,
+                ItemId = "new_Item_ID",
+                AssociatedObjectId =  "new_object_Id",
+                AssociatedObjectType = entity.AssociationType,
+            };
+            // Act
+            
+            await associationServiceMock.UpdateAssociationsAsync(new[] {productAssociation});
+            // Assert
+            Assert.Equal(productAssociation.ItemId, entity.ItemId);
```
Comment (yecli) [inline tests/VirtoCommerce.CatalogModule.Tests/ProductAssociationTests.cs:60]: Probably here you need to check that Add was called. And probably you are checking not adding here, but update entity with the same Id - so you need to ensure that Add was not called, and ItemId and AssociatedObjectId were changed.
Also probably the correct name of the method is `UpdateAssociationsAsync_UpdateNotTransientAssociation_Changed`

File: tests/VirtoCommerce.CatalogModule.Tests/ProductAssociationTests.cs [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+                AssociatedObjectId = entity.AssociatedItemId,
+                AssociatedObjectType = entity.AssociationType,
+                Type = entity.AssociationType,
+                Priority = 10
+            };
+            // Act
+
+            await associationServiceMock.UpdateAssociationsAsync(new[] { productAssociation });
+            // Assert
+            Assert.Equal(productAssociation.Priority, entity.Priority);
```
Comment (yecli) [inline tests/VirtoCommerce.CatalogModule.Tests/ProductAssociationTests.cs:83]: The same comment as before - need to check that repository.Add was not called, and change the name - it is not Add scenario

File: tests/VirtoCommerce.CatalogModule.Tests/ProductAssociationTests.cs [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+            // Act
+            await associationServiceMock.UpdateAssociationsAsync(new[] { productAssociation });
+            // Assert
+            _catalogRepositoryMock.Verify(x => x.Add(It.Is<dataModel.AssociationEntity>(q => q.AssociatedItemId == "NEW_AssociatedItemId")), Times.Once);
+        }
+
+
+        [Theory]
+        [MemberData(nameof(ValidEntities))]
+        public async Task SaveExistedAssociation(dataModel.AssociationEntity association)
```
Comment (yecli) [inline tests/VirtoCommerce.CatalogModule.Tests/ProductAssociationTests.cs:111]: SaveExist**ing**Association

File: tests/VirtoCommerce.CatalogModule.Tests/ProductAssociationTests.cs [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+        {
+            get
+            {
+                yield return new dataModel.ItemEntity { Id = "prod-1", CategoryId = "cat-1" };
+                yield return new dataModel.ItemEntity { Id = "prod-2", CategoryId = "cat-1-2" };
+                yield return new dataModel.ItemEntity { Id = "prod-3", CategoryId = "cat-1-2-3" };
+                yield return new dataModel.ItemEntity { Id = "prod-4", CategoryId = "cat-2" };
+                yield return new dataModel.ItemEntity { Id = "prod-5", CategoryId = "cat-3" };
+            }
+        }
```
Comment (yecli) [inline tests/VirtoCommerce.CatalogModule.Tests/ProductAssociationTests.cs:364]: Are these used?

File: tests/VirtoCommerce.CatalogModule.Tests/ProductAssociationTests.cs [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+                },
+
+            },
+
+
+        };
+
+        [Theory]
+        [MemberData(nameof(EqualEntities))]
+        public void EqualAssociation(dataModel.AssociationEntity x, dataModel.AssociationEntity y)
```
Comment (yecli) [inline tests/VirtoCommerce.CatalogModule.Tests/ProductAssociationTests.cs:274]: AssociationEntityComparer_EqualAssociations_Equals

File: tests/VirtoCommerce.CatalogModule.Tests/ProductAssociationTests.cs [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+            var result = new dataModel.AssociationEntityComparer().Equals(x, y);
+
+            Assert.False(result);
+        }
+
+
+
+
+
+
```
Comment (yecli) [inline tests/VirtoCommerce.CatalogModule.Tests/ProductAssociationTests.cs:336]: Please remove

**Pull request merged**